packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2705,9 +2705,9 @@ export class KeyringController extends BaseController<
     callback: MutuallyExclusiveCallback<Result>,
   ): Promise<Result> {
     return this.#withRollback(async ({ releaseLock }) => {
-      const oldState = await this.#getSessionState();
+      const oldState = JSON.stringify(await this.#getSessionState());
       const callbackResult = await callback({ releaseLock });
-      const newState = await this.#getSessionState();
+      const newState = JSON.stringify(await this.#getSessionState());
 
       // State is committed only if the operation is successful and need to trigger a vault update.
       if (!isEqual(oldState, newState)) {
```

Comment:
Nit: What do you think about updating this to, just to preserve the context around why we’re making this change?
```suggestion
      const oldState = await this.#getSessionState();
      const callbackResult = await callback({ releaseLock });
      const newState = await this.#getSessionState();

      // State is committed only if the operation is successful and need to trigger a vault update.
      // Use JSON.stringify comparison to handle cases where keyrings return shallow copies
      // of their internal state, which would cause isEqual to return false even when the
      // serialized content is identical.
      if (JSON.stringify(oldState) !== JSON.stringify(newState)) {

Reply:
I initially committed the suggestion, but I may have to revert it.
The reason I was stringifying the states before the callback execution is that doing it later would be "too late": by the time we compare the states, they will both be changed.

Reply:
Indeed, I missed that part!

Reply:
The joy of having shared references 🥹